src/Workspaces/Core/Portable/Workspace/Solution/DocumentInfo.cs

```diff
@@ -125,6 +113,18 @@ internal DocumentInfo(DocumentAttributes attributes, TextLoader? loader, IDocume
             return new DocumentInfo(newAttributes, newLoader, newDocumentServiceProvider);
         }
 
+        /// <summary>
+        /// A loader that can retrieve the document text.
+        /// </summary>
+        public TextLoader? TextLoader
+            => _textLoader is NullTextLoader ? null : _textLoader;
+
+        /// <summary>
+        /// Algorithm used for calculating the document content checksum.
+        /// </summary>
+        internal SourceHashAlgorithm ChecksumAlgorithm
+            => _textLoader.ChecksumAlgorithm;
```

Comment:
What's the intended use of this? The project has a checksum algorithm and so does the document -- how are we reconciling?

Reply:
Similarly to ParseOptions. The one that matters is the document's value. The project's value is there to store the default for documents that are being added.

Reply:
@tmat I get that we have to shuttle the value the DocumentState eventually, but a DocumentInfo doesn't hold the parse options -- that's held by the state directly. Traditionally the DocumentInfo is just the stuff that's required to initialize it, but it get accidentally coopted as a way to also hold other state. Is the expectation that this is just ignored (or copied from the project info)? Because if this is just fallout from the abuse by the state of the Info object that's something that we can clean up separately. It's caused confusion before especilly for ProjectState/ProjectInfo.

Reply:
There is no change re holding state here. It already held on TextLoader. The checksum algorithm is a property of the TextLoader.

Reply:
That said seems like this property is not used, other then in tests. I'll remove it.
Also, if we pass the checksum alg to LoadTextAndVersionAsync the state will be gone.

Reply:
Ah, never mind. FAR was lying. The property is indeed necessary. It's used in TextDocumentState constructor.